### PR TITLE
[3.10] bpo-44490: Add __parameters__ and __getitem__ to types.Union (GH-26980)

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Co-authored-by: Ken Jin <28750310+Fidget-Spinner@users.noreply.github.com>
Co-authored-by: Guido van Rossum <gvanrossum@gmail.com>.
(cherry picked from commit c45fa1a5d9b419cf13ad4b5a7cb453956495b83e)

Co-authored-by: Yurii Karabas <1998uriyyo@gmail.com>

https://bugs.python.org/issue44490

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Since you are backporting this, I assume it means you approve the original? One of the [conditions by Pablo](https://github.com/python/cpython/pull/26980#issuecomment-874284261) is that 2 core devs need to approve it, and we already have Guido's approval.

**Comment by serhiy-storchaka** [comment]
It includes changes made after merging the original PR.

There are two bugs left, but I am going to fix them soon.

**Comment by Fidget-Spinner** [comment]
LGTM but I don't understand why ``make check-abidump`` is failing in the CI:
```
2 Removed functions:

  'function PyObject* _Py_make_parameters(PyObject*)'    {_Py_make_parameters}
  'function PyObject* _Py_subs_parameters(PyObject*, PyObject*, PyObject*, PyObject*)'    {_Py_subs_parameters}

```
Also let's wait for Pablo :).

**Comment by pablogsal** [comment]
> LGTM but I don't understand why ``make check-abidump`` is failing in the CI:
> ```
> 2 Removed functions:
> 
>   'function PyObject* _Py_make_parameters(PyObject*)'    {_Py_make_parameters}
>   'function PyObject* _Py_subs_parameters(PyObject*, PyObject*, PyObject*, PyObject*)'    {_Py_subs_parameters}
> 
> ```
> Also let's wait for Pablo :).

I am a bit uncomfortable backporting 100 lines of C for a new feature bit I will allow based on:

* The feature is for a new type so there is not s very high risk of backwards incompatibility.
* The PR is authored by @serhiy-storchaka :)

@serhiy-storchaka I saw that you are preparing several bugfixes and cleanups for this type. Please, merge these ASAP as we are very close the the release candidate and I want them to be tested as much as possible.

Thanks!

**Review by pablogsal — APPROVED** [review]

**Comment by Fidget-Spinner** [comment]
> But please, understand that we cannot backport new features in general, specially so close to the release candidate, so don't take this as the norm :)

@pablogsal I understand that the last beta is generally the cut-off for new features, and that this is a rare exception by you. Thanks for allowing it! I'll try my best to ensure typing-related stuff doesn't cause you any more release headaches (especially on a weekend) for the 3.11 cycle.

One unfortunate cause for things like these is that it seems that not many people actually test the betas until the last one, or when the release candidates arrive. So bugs/missing features get caught very late :(.

**Comment by pablogsal** [comment]
> One unfortunate cause for things like these is that it seems that not many people actually test the betas until the last one, or when the release candidates arrive. So bugs/missing features get caught very late :(.

They actually test it, but not so much for new features. The 90% of the tests they do is based on compatibility with existing code, and they don't stress test that much the new stuff

**Comment by Fidget-Spinner** [comment]
> They actually test it, but not so much for new features. The 90% of the tests they do is based on compatibility with existing code, and they don't stress test that much the new stuff

Hmm I realised my original wording sounds kinda unfair on large packages that do test the betas. Yeah I meant just for the new features. (Sorry, no harm intended). I'm well aware third party package like Django have been testing the betas since day 1 :).

**Comment by pablogsal** [comment]
> > They actually test it, but not so much for new features. The 90% of the tests they do is based on compatibility with existing code, and they don't stress test that much the new stuff
> 
> Hmm I realised my original wording sounds kinda unfair on large packages that do test the betas. Yeah I meant just for the new features. (Sorry, no harm intended). I'm well aware third party packages like Django have been testing the betas since day 1 :).

No problem, I understood your intent ;)

**Comment by gvanrossum** [comment]
> I understand that the last beta is generally the cut-off for new features

Actually the *first* beta is the cut-off for new features.

**Comment by Fidget-Spinner** [comment]
> Actually the _first_ beta is the cut-off for new features.

Oops right thanks, I blame timezones for my lapse in memory :P.

In any case, the windows x64 tests are unrelated `test_asyncio` failures. The `make check-abidump` failure doesn't make sense to me. It's saying we removed those functions from the stable ABI, but those aren't in 3.10 to begin with?

**Comment by pablogsal** [comment]
> > Actually the _first_ beta is the cut-off for new features.
> 
> Oops right thanks, I blame timezones for my lapse in memory :P.
> 
> In any case, the windows x64 tests are unrelated `test_asyncio` failures. The `make check-abidump` failure doesn't make sense to me. It's saying we removed those functions from the stable ABI, but those aren't in 3.10 to begin with?

If that check if failing it means that this PR is affecting the ABI in any way, it doesn't distinguish the public or private one

File: Include/internal/pycore_unionobject.h
Comment (Fidget-Spinner) [inline Include/internal/pycore_unionobject.h]: Thanks to Pablo's comments above, I finally understand what the check is doing. Seems that release branches have a frozen generated list of exported symbols which are tested against. So there a few solutions:

1. Follow instructions at https://bugs.python.org/issue43725#msg390213 and regen the exported symbols. Though this looks somewhat time consuming.
2. Add ignorefile.
3. Remove the `PYAPI_FUNC` (I like this most):
```suggestion
PyObject *_Py_subs_parameters(PyObject *, PyObject *, PyObject *, PyObject *);
PyObject *_Py_make_parameters(PyObject *);
```
I don't have the proper environment to test that option 3 will really make the test pass. But I think it should.

**Comment by gvanrossum** [comment]
Yeah, the Py_API is for exported names, not for purely internal functions.

**Review by gvanrossum — APPROVED** [review]

**Pull request merged**
